src/aaudio/AudioStreamAAudio.h

```diff
@@ -98,24 +98,53 @@ class AudioStreamAAudio : public AudioStream {
 
     void onErrorInThread(AAudioStream *stream, Result error);
 
-
-    void *getUnderlyingStream() const override {
-        return mAAudioStream.load();
-    }
+    bool usesMMap() override;
 
 protected:
     void updateFramesRead() override;
     void updateFramesWritten() override;
 
+    static AAudioLoader *mLibLoader;
+
 private:
 
-    bool                 isMMapUsed();
+    /** Define a class that wraps a 'C' style AAudioStream pointer.
+     * We can then use a std:shared_ptr to manage the ownership of that pointer.
+     */
+    class   AAudioStreamOwner {
+    public:
+
+        // Create the stream in the constructor to avoid race conditions.
+        AAudioStreamOwner(AAudioStreamBuilder *aaudioBuilder) {
+            AAudioStream *stream = nullptr;
+            mOpenResult = mLibLoader->builder_openStream(aaudioBuilder, &stream);
+            mAAudioStream.store(stream);
+        }
+
+        ~AAudioStreamOwner() {
+            if (mAAudioStream) {
+                // Calling stream_close() will free memory associated with the stream.
+                AudioStreamAAudio::mLibLoader->stream_close(mAAudioStream.load());
+                mAAudioStream.store(nullptr);
+            }
+        }
+        aaudio_result_t getOpenResult() {
+            return mOpenResult;
+        }
+
+        AAudioStream *getStream() {
+            return mAAudioStream.load();
+        }
+
+    private:
+        aaudio_result_t mOpenResult = AAUDIO_OK;
```

Comment:
I think it would be better not to have 'mOpenResult' here. Instead, provide a static factory method which returns a shared_ptr&lt;AAudioStreamOwner&gt;, and fills out the result as an out parameter. Make AAudioStreamOwner's constructor private.